File: modules/biovision-facephys/ios/TFLiteInterpreter.swift
import Foundation
import TensorFlowLiteC

enum FacePhysError: LocalizedError {
  case missingResource(String)
  case modelLoadFailed(String)
  case tensorNotFound(String)
  case inferenceFailed(String)

  var errorDescription: String? {
    switch self {
    case .missingResource(let name): return "FacePhys resource \(name) is missing from the app bundle."
    case .modelLoadFailed(let name): return "FacePhys model \(name) could not be loaded."
    case .tensorNotFound(let name): return "FacePhys tensor \(name) was not found."
    case .inferenceFailed(let name): return "FacePhys model \(name) failed to run."
    }
  }
}

/// Thin wrapper over the TensorFlow Lite C API. Tensors are addressed by name,
/// because the browser runtime and native TFLite enumerate them in different orders.
final class TFLiteInterpreter {
  private let name: String
  private let model: OpaquePointer
  private let options: OpaquePointer
  private let interpreter: OpaquePointer
  private let delegate: UnsafeMutablePointer<TfLiteDelegate>?
  private var inputIndexByName: [String: Int32] = [:]
  private var outputIndexByName: [String: Int32] = [:]

  init(resource: String, threads: Int32 = 2) throws {
    name = resource
    guard let url = FacePhysResources.url(for: resource) else { throw FacePhysError.missingResource(resource) }
    guard let model = TfLiteModelCreateFromFile(url.path) else { throw FacePhysError.modelLoadFailed(resource) }
    guard let options = TfLiteInterpreterOptionsCreate() else {
      TfLiteModelDelete(model)
      throw FacePhysError.modelLoadFailed(resource)
    }
    TfLiteInterpreterOptionsSetNumThreads(options, threads)
    // XNNPACK is TFLite's optimized CPU backend; unsupported ops fall back to the default kernels.
    var delegateOptions = TfLiteXNNPackDelegateOptionsDefault()
    delegateOptions.num_threads = threads
    let delegate = TfLiteXNNPackDelegateCreate(&delegateOptions)
    if let delegate { TfLiteInterpreterOptionsAddDelegate(options, delegate) }
    guard let interpreter = TfLiteInterpreterCreate(model, options),
          TfLiteInterpreterAllocateTensors(interpreter) == kTfLiteOk else {
      TfLiteInterpreterOptionsDelete(options)
      if let delegate { TfLiteXNNPackDelegateDelete(delegate) }
      TfLiteModelDelete(model)
      throw FacePhysError.modelLoadFailed(resource)
    }
    self.delegate = delegate
    self.model = model
    self.options = options
    self.interpreter = interpreter
    for index in 0..<TfLiteInterpreterGetInputTensorCount(interpreter) {
      if let tensor = TfLiteInterpreterGetInputTensor(interpreter, index), let raw = TfLiteTensorName(tensor) {
        inputIndexByName[String(cString: raw)] = index
      }
    }
    for index in 0..<TfLiteInterpreterGetOutputTensorCount(interpreter) {
      if let tensor = TfLiteInterpreterGetOutputTensor(interpreter, index), let raw = TfLiteTensorName(tensor) {
        outputIndexByName[String(cString: raw)] = index
      }
    }
  }

  deinit {
    TfLiteInterpreterDelete(interpreter)
    if let delegate { TfLiteXNNPackDelegateDelete(delegate) }
    TfLiteInterpreterOptionsDelete(options)
    TfLiteModelDelete(model)
  }

  var inputNames: [String] { Array(inputIndexByName.keys) }

  /// Writable float32 storage of a named input tensor. Stable after allocation.
  func input(_ tensorName: String) throws -> UnsafeMutableBufferPointer<Float32> {
    guard let index = inputIndexByName[tensorName],
          let tensor = TfLiteInterpreterGetInputTensor(interpreter, index),
          let data = TfLiteTensorData(tensor) else { throw FacePhysError.tensorNotFound(tensorName) }
    let count = TfLiteTensorByteSize(tensor) / MemoryLayout<Float32>.stride
    return UnsafeMutableBufferPointer(start: data.assumingMemoryBound(to: Float32.self), count: count)
  }

  func output(_ tensorName: String) throws -> UnsafeBufferPointer<Float32> {
    guard let index = outputIndexByName[tensorName],
          let tensor = TfLiteInterpreterGetOutputTensor(interpreter, index),
          let data = TfLiteTensorData(tensor) else { throw FacePhysError.tensorNotFound(tensorName) }
    let count = TfLiteTensorByteSize(tensor) / MemoryLayout<Float32>.stride
    return UnsafeBufferPointer(start: UnsafeRawPointer(data).assumingMemoryBound(to: Float32.self), count: count)
  }

  func invoke() throws {
    guard TfLiteInterpreterInvoke(interpreter) == kTfLiteOk else { throw FacePhysError.inferenceFailed(name) }
  }
}

enum FacePhysResources {
  private static let bundle: Bundle? = {
    let owner = Bundle(for: TFLiteInterpreter.self)
    guard let url = owner.url(forResource: "BioVisionFacePhysModels", withExtension: "bundle") else { return nil }
    return Bundle(url: url)
  }()

  static func url(for resource: String) -> URL? {
    let name = (resource as NSString).deletingPathExtension
    let ext = (resource as NSString).pathExtension
    return bundle?.url(forResource: name, withExtension: ext)
  }
}

File: modules/biovision-facephys/android/src/main/java/com/biovision/facephys/TFLiteInterpreter.kt
package com.biovision.facephys

import android.content.Context
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.channels.FileChannel
import org.tensorflow.lite.Interpreter

class FacePhysException(message: String) : RuntimeException(message)

/**
 * Thin wrapper over the TensorFlow Lite Java API, mirroring ios/TFLiteInterpreter.swift.
 * Tensors are addressed by name, because the browser runtime and native TFLite enumerate
 * them in different orders.
 *
 * iOS writes straight into tensor memory through the C API. The Java API exposes no such
 * pointer, so every input and output is backed by a direct ByteBuffer that is allocated
 * once and reused: the FacePhys engine relies on input contents persisting across frames
 * (the recurrent state tensors), which a per-call array would lose.
 */
class TFLiteInterpreter(context: Context, private val name: String, threads: Int = 2) {
  private val interpreter: Interpreter
  private val inputIndexByName = mutableMapOf<String, Int>()
  private val outputIndexByName = mutableMapOf<String, Int>()
  private val inputBuffers = mutableMapOf<Int, ByteBuffer>()
  private val outputBuffers = mutableMapOf<Int, ByteBuffer>()
  private val inputViews = mutableMapOf<Int, FloatBuffer>()
  private val outputViews = mutableMapOf<Int, FloatBuffer>()

  init {
    val model = try {
      loadAsset(context, name)
    } catch (error: Exception) {
      throw FacePhysException("FacePhys model $name could not be loaded: ${error.message}")
    }
    val options = Interpreter.Options().apply {
      numThreads = threads
      // XNNPACK is TFLite's optimized CPU backend; unsupported ops fall back to the
      // default kernels, matching the iOS delegate setup.
      setUseXNNPACK(true)
    }
    interpreter = Interpreter(model, options)
    interpreter.allocateTensors()

    for (index in 0 until interpreter.inputTensorCount) {
      val tensor = interpreter.getInputTensor(index)
      inputIndexByName[tensor.name()] = index
      val buffer = ByteBuffer.allocateDirect(tensor.numBytes()).order(ByteOrder.nativeOrder())
      inputBuffers[index] = buffer
      inputViews[index] = buffer.asFloatBuffer()
    }
    for (index in 0 until interpreter.outputTensorCount) {
      val tensor = interpreter.getOutputTensor(index)
      outputIndexByName[tensor.name()] = index
      val buffer = ByteBuffer.allocateDirect(tensor.numBytes()).order(ByteOrder.nativeOrder())
      outputBuffers[index] = buffer
      outputViews[index] = buffer.asFloatBuffer()
    }
  }

  /**
   * Models live in the library's assets. They are stored uncompressed (see `noCompress` in
   * build.gradle) so they can be memory-mapped instead of copied onto the heap.
   */
  private fun loadAsset(context: Context, asset: String): ByteBuffer {
    val descriptor = context.assets.openFd(asset)
    FileInputStream(descriptor.fileDescriptor).use { stream ->
      return stream.channel.map(
        FileChannel.MapMode.READ_ONLY,
        descriptor.startOffset,
        descriptor.declaredLength,
      ).order(ByteOrder.nativeOrder())
    }
  }

  val inputNames: List<String> get() = inputIndexByName.keys.toList()

  /** Writable float view of a named input tensor; contents persist between invocations. */
  fun input(tensorName: String): FloatBuffer {
    val index = inputIndexByName[tensorName]
      ?: throw FacePhysException("FacePhys tensor $tensorName was not found.")
    return inputViews.getValue(index)
  }

  /** Float view of a named output tensor, refreshed by [invoke]. */
  fun output(tensorName: String): FloatBuffer {
    val index = outputIndexByName[tensorName]
      ?: throw FacePhysException("FacePhys tensor $tensorName was not found.")
    return outputViews.getValue(index)
  }

  /** Runs the graph over the persistent buffers. */
  fun invoke() {
    val inputs = arrayOfNulls<Any>(interpreter.inputTensorCount)
    for ((index, buffer) in inputBuffers) {
      buffer.rewind()
      inputs[index] = buffer
    }
    val outputs = mutableMapOf<Int, Any>()
    for ((index, buffer) in outputBuffers) {
      buffer.rewind()
      outputs[index] = buffer
    }
    try {
      interpreter.runForMultipleInputsOutputs(inputs, outputs)
    } catch (error: Exception) {
      throw FacePhysException("FacePhys model $name failed to run: ${error.message}")
    }
  }

  fun close() = interpreter.close()
}
